server/command.js

```diff
@@ -0,0 +1,243 @@
+  var hosts = [user, user.location, spec.dobj, spec.iobj];
+  for (var i = 0; i < hosts.length; i++) {
+    var host = hosts[i];
+    if (!host) {
+      continue;
+    }
+    // Check every verb on each object for a match.
+    var func, verb, dobj, prepstr, iobj;
+    for (var prop in host) {
+      func = host[prop];
+      if (typeof func === 'function' && (prepstr = func.prepstr) &&
+          (dobj = func.dobj) && (iobj = func.iobj && (verb = func.verb))) {
+        // This is a verb.
+        if (prepstr === 'any' || prepstr == spec.prepstr ||
+            (prepstr == 'none' && !prepstr)) {
+          if (dobj === 'any' || (dobj === 'this' && spec.dobj === host) ||
+              (dobj === 'none' && !dobj)) {
+            if (iobj === 'any' || (iobj === 'this' && spec.iobj === host) ||
+                (iobj === 'none' && !iobj)) {
+              if (spec.verb.search(verb) === 0) {
+                host[prop](spec);
+                return true;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return false;
+};
+
+$.utils.match = function(str, context) {
+  // Attempt to match a string with the environmental context.
+  // Context can be anything (including null) but is normally a place or a user.
+  // Returns an array of all matching objects.
+  str = str.trim();
+  // First, check for matches against universally accessible things.
+  var m = $(str);
+  if (m !== undefined) {
+    return [m];
+  }
+  // Second, check the context for one or more matches.
+  if (!(context instanceof $.physical)) {
+    return [];
+  }
+  var nameMatches = [];  // These should be Sets.
+  var aliasMatches = [];
+  var partialMatches = [];
+  function matchObj(obj) {
+    var m = $.utils.match.strength(str, obj);
+    m && (m === 1 ? partialMatches : (m === 2 ? aliasMatches : nameMatches))
+        .push(obj);
+  }
+  // Check the contextual object.
+  matchObj(context);
+  // Check the contextual object's contents.
+  var contents = context.contents();
+  for (var i = 0; i < contents.length; i++) {
+    matchObj(contents[i]);
+  }
+  var location = context.location;
+  if (context instanceof $.user && location instanceof $.physical) {
+    // Try the user's place next.
+    matchObj(location);
+    // Check the user's place's contents.
+    var contents = context.contents();
+    for (var i = 0; i < contents.length; i++) {
+      if (contents[i] !== context) {
+        matchObj(contents[i]);
+      }
+    }
+  }
+  // Return the highest level bin.
+  if (nameMatches.length) {
+    return nameMatches;
+  }
+  if (aliasMatches.length) {
+    return aliasMatches;
+  }
+  if (partialMatches.length) {
+    return partialMatches;
+  }
+  return [];
+};
+
+$.utils.match.strength = function(str, obj) {
+  // On a score of zero to three, how well does str match obj?
+  // 0: No match.
+  // 1: Partial name or alias.
+  // 2: Perfect alias match.
+  // 3: Perfect name match.
+  if (!str || !obj) {
```

Comment:
No problem rejecting empty strings, but as-is this code will crash if `str` is a number, or if `obj` has a `null` prototype (or indeed fails to have `.aliases` defined as an array).